NeedALight/magnus.py

```diff
@@ -0,0 +1,431 @@
+    xmin2 = np.array([-1, -1, 0, 0])
+    xmax2 = np.array([1, 1, 1, 1])
+
+    # Using cubature
+    test2, _err = cubature(func2, ndim2, fdim2, xmin2, xmax2, vectorized=False)
+
+    # Reshaping vector into Matrix and removing the convergence factors
+    J3_2 = test2.reshape(len(w), len(w)) / (0.01**3) / 0.1
+
+    return J3_1 + J3_2
+
+
+def Magnus3_Im(F, w):
+    """Generates the imaginary part of the third order Magnus correction J3
+
+    Args:
+        F (function): product of phase-matching function and pump pulse
+        w (array): list of frequencies
+
+    Returns:
+        (array): K3(w,w'): imagineray part of third order Magnus term
+    """
+
+    # Defining the K3 function for cubature. Extra factor for convergence.
+
+    K3 = lambda x, y, z, ws, wi: (0.01**3) * (
+        F(ws, y, ws + y + z) * np.conj(F(x, y, x + y + z)) * F(x, wi, wi + x)
+        + F(ws, x, ws + x) * F(y, wi, wi + y + z) * np.conj(F(y, x, x + y + z))
+    )
+
+    # We map integrands from (-inf,inf) to (-1,1).
+    # Principal values are symmetrized and evaluated from (0,inf)->(0,1).
+    def func3(x_array):
+        x = x_array[0]
+        y = x_array[1]
+        z = x_array[2]
+        return (
+            (
+                K3(x / (1 - x**2), y / (1 - y**2), z / (1 - z), w, w[:, np.newaxis])
+                - K3(x / (1 - x**2), y / (1 - y**2), -z / (1 - z), w, w[:, np.newaxis])
+            )
+            * ((1 + x**2) / ((1 - x**2) ** 2))
+            * ((1 + y**2) / ((1 - y**2) ** 2))
+            * (1 / (z * (1 - z)))
+        ).flatten()
+
+    # Integral properties
+    ndim3 = 3  # Number of variables we are integrating over
+    fdim3 = len(w) ** 2  # Dimension of the output/J3
+    xmin3 = np.array([-1, -1, 0])
+    xmax3 = np.array([1, 1, 1])
+
+    # Using cubature
+    test3, _err3 = cubature(
+        func3, ndim3, fdim3, xmin3, xmax3, adaptive="h", vectorized=False
+    )
+
+    # Reshaping vector into Matrix and removing the convergence factors
+    K3 = test3.reshape(len(w), len(w)) / (0.01**3)
+
+    return K3
+
+def Magnus1CW_data(ws,wi,k_o,k_e,k_pol,p_k,L):
+    """Generates first Magnus term for Continous-Wave model from given data set
+    
+    Args:
+        ws (array): signal frequencies of interest, interpolated from data given a range of wavelengths
+        wi (array): corresponding idler frequencies such that wi=wp_avg-ws
+        k_o (function): interpolating function generating wavevector given ordinary(signal) frequency
+        k_e (function): interpolating function generating wavevector given extra-ordinary(idler) frequency
+        k_pol (float): poling wavevector to induce conservation of momenta
+        p_k (float): wavevector of CW pump   
+        L (float): length of interaction region/crystal
+    
+    Returns:
+        J1(w): first order magnus term.
+    
+    """
+    J1 = np.sinc(L*(k_o(ws)+k_e(wi)+k_pol-p_k)/(2*np.pi))
+
+
+
+    return J1
+
+def Magnus3CW_data_Re(omega,ws,wi,k_o,k_e,k_pol,p_k,L):
```

Comment:
You probably want to pass black in this file.

Also, you want to make the scaling argument as an optional parameter in the functions. That way you don't need to change your tests but you still give freedom to the user without them having to go an open the file.